src/Type/Php/StrCaseFunctionsReturnTypeExtension.php

```diff
@@ -0,0 +1,89 @@
+<?php declare(strict_types = 1);
+
+namespace PHPStan\Type\Php;
+
+use PhpParser\Node\Expr\FuncCall;
+use PHPStan\Analyser\Scope;
+use PHPStan\Reflection\FunctionReflection;
+use PHPStan\Reflection\ParametersAcceptorSelector;
+use PHPStan\ShouldNotHappenException;
+use PHPStan\Type\Accessory\AccessoryNonEmptyStringType;
+use PHPStan\Type\Accessory\AccessoryNumericStringType;
+use PHPStan\Type\Constant\ConstantStringType;
+use PHPStan\Type\DynamicFunctionReturnTypeExtension;
+use PHPStan\Type\IntersectionType;
+use PHPStan\Type\StringType;
+use PHPStan\Type\Type;
+use PHPStan\Type\TypeUtils;
+use PHPStan\Type\UnionType;
+use function count;
+use function in_array;
+use function is_callable;
+
+class StrCaseFunctionsReturnTypeExtension implements DynamicFunctionReturnTypeExtension
+{
+
+	public function isFunctionSupported(FunctionReflection $functionReflection): bool
+	{
+		return in_array($functionReflection->getName(), [
+			'strtoupper',
+			'strtolower',
+			'mb_strtoupper',
+			'mb_strtolower',
+			'lcfirst',
+			'ucfirst',
+			'ucwords',
+		], true);
+	}
+
+	public function getTypeFromFunctionCall(
+		FunctionReflection $functionReflection,
+		FuncCall $functionCall,
+		Scope $scope,
+	): Type
+	{
+		$args = $functionCall->getArgs();
+		if (count($args) < 1) {
+			return ParametersAcceptorSelector::selectSingle($functionReflection->getVariants())->getReturnType();
+		}
+
+		$argType = $scope->getType($args[0]->value);
+		$fnName = $functionReflection->getName();
+		if (!is_callable($fnName)) {
+			throw new ShouldNotHappenException();
+		}
+
+		if (count($args) === 1) {
+			$constantStrings = TypeUtils::getConstantStrings($argType);
+			if (count($constantStrings) > 0) {
+				$strings = [];
+
+				foreach ($constantStrings as $constantString) {
+					$strings[] = new ConstantStringType($fnName($constantString->getValue()));
+				}
+
+				if (count($strings) === 1) {
+					return $strings[0];
+				}
+				return new UnionType($strings);
```

Comment:
You don't need the if here. Using `TypeCombinator::union()` is more correct. You can have an edge case of calling `lcfirst` of `'foo'|'Foo'` and the current code would result in `'foo'|'foo'` which is bad :)

Reply:
oh wow... how do you come up with such ideas...? well spotted!

Reply:
I guess the experience from doing this every day for the past 6 years 😅